add ApplyPlanningSceneService capability

v4hn opened the pull request:
See https://groups.google.com/forum/#!topic/moveit-users/D7MWZEUSKLc
for a longer discussion on why this makes sense to have around.

The only thing I'm unsure about is whether the service should be
`~`-namespaced or not. There might be multiple monitored scenes
around, so I moved it to the namespace. On the other hand,
the ClearOctomap service is _not_ namespaced...

v4hn committed:
monitor: make newPlanningSceneMessage public

This is required for capabilities to update the planning scene.

ABI-compatible.

v4hn committed:
add apply_planning_scene capability

This new capability allows to apply changes to a monitored planning
scene and *blocks* until the changes are applied. This is meant to
replace the quasi-standard pattern:
```
planning_scene_interface.addCollisionObjects(...)
sleep(2.0)
group.pick("object")
```
by
```
ros::ServiceClient client = n.serviceClient<moveit_msgs::ApplyPlanningScene>("apply_planning_scene");
client.call(...)
group.pick("object")
```

This makes it much more convenient to add&interact with objects
without useless and arbitrarily long sleeps to ensure planning scene
updates succeeded.

v4hn commented:
This obviously depends on https://github.com/ros-planning/moveit_msgs/pull/21

davetcoleman commented on move_group/src/default_capabilities/apply_planning_scene_service_capability.h:
```diff
@@ -0,0 +1,63 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Michael 'v4hn' Goerner */
```
im a big fan of descriptions in header files for what the class does....

v4hn replied:
On Mon, Jun 13, 2016 at 03:41:28PM -0700, Dave Coleman wrote:

> im a big fan of descriptions in header files for what the class does....

Now you're being unfair. :)
I totally agree, but none of the other capabilities include a description
and I wouldn't even know what to describe here, as it's pretty self-documenting.
"ApplyPlanningSceneService : MoveGroupCapability is a MoveGroupCapability
that provides a service to apply a given PlanningScene" is just overly redundant.

davetcoleman replied:
You are exactly right about not having enough documentation throughout the code base and I know the others lack it. How about:
"Provides ability to update the shared planning scene with a remote blocking call"

davetcoleman commented on planning/planning_scene_monitor/include/moveit/planning_scene_monitor/planning_scene_monitor.h:
```diff
@@ -478,9 +481,6 @@ class PlanningSceneMonitor : private boost::noncopyable
   void newPlanningSceneCallback(const moveit_msgs::PlanningSceneConstPtr &scene);
 
-  // Called to update the planning scene with a new message.
-  void newPlanningSceneMessage(const moveit_msgs::PlanningScene& scene);
```
just a guess, but wouldn't this break ABI too? I don't care about ABI, actually, but maybe we should go ahead and make it return bool while we're at it (doesn't break API).

v4hn replied:
On Mon, Jun 13, 2016 at 03:42:43PM -0700, Dave Coleman wrote:

> just a guess, but wouldn't this break ABI too?

No, changing visibility private -> public does not break ABI.
As it turns out, in this case changing the `void` return
to `bool` also is ABI compliant as the function was private before.
I'll prepare the changes.

v4hn committed:
simplify newPlanningSceneMessage

v4hn committed:
apply planning scene: use newly exposed success value of newPlanningSceneMessage

v4hn commented:
@davetcoleman: now this request also spans 4 repos :-)
Full +1 to merge most moveit repositories...

I do care quite a lot about a compatible ABI within ros distros until we add sonames to our libraries to explicitly mark updates as ABI-breaking.
As @mikeferguson pointed out last week it's pretty horrible to debug random segfaults due to ABI changes without sonames. Last year I spent more than 12 hours debugging one of these issues in the PCL, although it was a particularly horrible one.

Any comments on the question of namespacing?

davetcoleman commented on planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -476,73 +476,77 @@ void planning_scene_monitor::PlanningSceneMonitor::clearOctomap()
+  if (!scene_)
+    return false;
+
+  bool ret;
```
result

davetcoleman commented on planning/planning_scene_monitor/src/planning_scene_monitor.cpp:
```diff
@@ -476,73 +476,77 @@ void planning_scene_monitor::PlanningSceneMonitor::clearOctomap()
+
+  bool ret;
+
+  SceneUpdateType upd = UPDATE_SCENE;
```
again, full var names...

davetcoleman commented:
i don't have an opinion on the namespacing, other than we should do it as the other services have been done. i don't know of any current uses cases needing more than one planning scene

v4hn commented:
Ok, I renamed all the `result` variables,
added the comment you proposed, and decided to move
the service away from `~`. Most of the other services
don't use it so no reason to start with it for now.

davetcoleman commented:
I'll merge these as soon as @mikeferguson gives me a +1 on the moveit_msgs change. Until then, could you add a section to the documentation [here](http://docs.ros.org/indigo/api/pr2_moveit_tutorials/html/planning/src/doc/planning_scene_ros_api_tutorial.html) mentioning this new way to update the PS? It doesn't need to be much.

mikeferguson commented:
@davetcoleman -- we need to merge the moveit_msgs and then release it as Debs so that this can properly build, else the CI will be broken

davetcoleman commented:
just restarted build after merging msgs and core
